pandera/backends/pandas/container.py

```diff
@@ -47,53 +47,89 @@ def validate(
+                check_obj = check_obj.pandera.add_schema(schema)
 
-        column_info = self.collect_column_info(check_obj, schema, lazy)
+            column_info = self.collect_column_info(check_obj, schema, lazy)
 
-        # collect schema components
-        components = self.collect_schema_components(
-            check_obj, schema, column_info
-        )
+            # collect schema components
+            components = self.collect_schema_components(
+                check_obj, schema, column_info
+            )
 
-        core_parsers: List[Tuple[Callable[..., Any], Tuple[Any, ...]]] = [
-            (self.strict_filter_columns, (schema, column_info)),
-            (self.coerce_dtype, (schema,)),
-        ]
+            core_parsers: List[Tuple[Callable[..., Any], Tuple[Any, ...]]] = [
+                (self.strict_filter_columns, (schema, column_info)),
+                (self.coerce_dtype, (schema,)),
+            ]
 
-        for parser, args in core_parsers:
-            try:
-                check_obj = parser(check_obj, *args)
-            except SchemaError as exc:
-                error_handler.collect_error(exc.reason_code, exc)
-            except SchemaErrors as exc:
-                error_handler.collect_errors(exc)
-
-        # subsample the check object if head, tail, or sample are specified
-        sample = self.subsample(check_obj, head, tail, sample, random_state)
-
-        # check the container metadata, e.g. field names
-        core_checks = [
-            (self.check_column_names_are_unique, (check_obj, schema)),
-            (self.check_column_presence, (check_obj, schema, column_info)),
-            (self.check_column_values_are_unique, (sample, schema)),
-            (self.run_schema_component_checks, (sample, components, lazy)),
-            (self.run_checks, (sample, schema)),
-        ]
+            for parser, args in core_parsers:
+                try:
+                    check_obj = parser(check_obj, *args)
+                except SchemaError as exc:
+                    error_handler.collect_error(exc.reason_code, exc)
+                except SchemaErrors as exc:
+                    error_handler.collect_errors(exc)
+
+            # subsample the check object if head, tail, or sample are specified
+            sample = self.subsample(
+                check_obj, head, tail, sample, random_state
+            )
+
+            # check the container metadata, e.g. field names
+            core_checks = [
+                (self.check_column_names_are_unique, (check_obj, schema)),
+                (self.check_column_presence, (check_obj, schema, column_info)),
+                (self.check_column_values_are_unique, (sample, schema)),
+                (self.run_schema_component_checks, (sample, components, lazy)),
+                (self.run_checks, (sample, schema)),
+            ]
+
+            # run the checks
+            error_handler = self.__run_checks_and_handle_errors(
```

Comment:
@cosmicBboy how do you feel about name mangling? Can just as easily swap to a single `_`

Reply:
let's do a little refactoring here:

- if we move the core checks to its own `run_core_checks` method (same for `run_core_parsers`, we clean this up a bit (I'm not familiar with the name mangling, but I don't think it's needed here: exposing a public method for running the core checks makes sense to me.
- we don't need the `try... except` here, since we have an error_handler object that we can use to check if there are schema errors.
- The `drop_invalid` logic can live underneath `if error_handler.collected_errors:` and return the check object.
- Probably worth creating a `drop_invalid_data` method to encapsulate the dropping logic. I was also thinking about how to handle metadata-level failures (e.g. wrong data type): should those columns be dropped to?

Reply:
Thanks for the tips! I have updated the PR accordingly. 

On the last point though, currently the PR would remove the rows in the `check_obj` where the index was a failure cases, such as for the reason of being an incorrect data type. If the whole column was incorrect, then the whole `check_obj` would be wiped 😱

Reply:
I think removing invalid columns is slightly different than dropping invalid rows, and could make more sense to treat them as distinct cases with distinct apis?